c/src/lib/libbsp/arm/shared/stm32fxxxx/spi/hal-spi.c

```diff
@@ -0,0 +1,507 @@
+/**
+ * @file hal-spi.c
+ *
+ * @ingroup spi
+ *
+ * @brief SPI driver for the STM32xxxx series processors. Provides a
+ * SPI driver that registers with cpukit/dev/spi as a node in the
+ * filesystem.
+ *
+ */
+
+/*
+ * Copyright (c) 2015 Vecna Technologies, Inc.
+ *
+ * The license and distribution terms for this file may be
+ * found in the file LICENSE in this distribution or at
+ * http://www.rtems.com/license/LICENSE.
+ */
+
+//================== STMF32 Support Functions =================================
+#include <rtems.h>
+#include <hal-utils.h>
+#include <stm32f-processor-specific.h>
+#include <bspopts.h>
+
+#include stm_processor_header(TARGET_STM_PROCESSOR_PREFIX)
+#include stm_header(TARGET_STM_PROCESSOR_PREFIX, dma)
+#include stm_header(TARGET_STM_PROCESSOR_PREFIX, spi)
+#include stm_header(TARGET_STM_PROCESSOR_PREFIX, rcc)
+#include stm_header(TARGET_STM_PROCESSOR_PREFIX, gpio)
+
+#include <hal-spi.h>
+#include <dev/spi/spi.h>
+
+#include <math.h>
+#include <rtems/irq-extension.h>
+
+/**
+ * SPI Instance enabled / disabled
+ */
+bool SPI_Enable[MAX_SPI_INSTANCES] =
+{
+#ifndef STM32_ENABLE_SPI1
+      false,
+#else
+      STM32_ENABLE_SPI1,
+#endif
+#ifndef STM32_ENABLE_SPI2
+      false,
+#else
+      STM32_ENABLE_SPI2,
+#endif
+#ifndef STM32_ENABLE_SPI3
+      false,
+#else
+      STM32_ENABLE_SPI3,
+#endif
+#ifndef STM32_ENABLE_SPI4
+      false,
+#else
+      STM32_ENABLE_SPI4,
+#endif
+#ifndef STM32_ENABLE_SPI5
+      false,
+#else
+      STM32_ENABLE_SPI5,
+#endif
+#ifndef STM32_ENABLE_SPI6
+      false,
+#else
+      STM32_ENABLE_SPI6,
+#endif
+};
+
+/**
+ * SPI pin configuration for each instances.
+ * Pin and port details to be populated in configure.ac file
+ */
```

Comment:
The #ifdefs allows us to selectively enable/disable the SPI instances. There are two reasons why using a global array wouldn't work out: First, we would like to control most of the peripheral config through configure.ac file during build and not modify the code. Second, to make the code generic to F4 and F7- some of the macros and instances (SPI4-6) are not supported in the F4 and we can remove that part of the code for F4 by undefining the macros.

This makes the code depend on a lot many #ifdefs by doing this, so limiting to only part of the code which needs to be stripped out for F4. But yes, I'm, sure there are better way to do it. :)